homeassistant/components/deconz.py

```diff
@@ -0,0 +1,134 @@
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _shutdown)
+
+    @asyncio.coroutine
+    def generate_api_key(call):
+        """Generate API key needed to communicate with Deconz.
+
+        Store API key in deconz.conf.
+        """
+        from pydeconz.utils import get_api_key
+        deconz_config[CONF_USERNAME] = call.data.get(CONF_USERNAME, 'delight')
+        deconz_config[CONF_PASSWORD] = call.data.get(CONF_PASSWORD, 'delight')
+        api_key = yield from get_api_key(hass.loop, **deconz_config)
+        if not save_json(
+                hass.config.path(CONFIG_FILE), {CONF_API_KEY: api_key}):
+            _LOGGER.error("Failed to save API key to %s", CONFIG_FILE)
+        deconz_config[CONF_API_KEY] = api_key
+        yield from _setup_deconz(hass, config, deconz_config)
+
+    if CONF_API_KEY in deconz_config:
+        yield from _setup_deconz(hass, config, deconz_config)
+    elif CONF_API_KEY in config_file:
+        deconz_config[CONF_API_KEY] = config_file[CONF_API_KEY]
+        yield from _setup_deconz(hass, config, deconz_config)
+    else:
+        hass.services.async_register(
+            DOMAIN, 'generate_api_key', generate_api_key)
+        _LOGGER.warning("Deconz needs API key to set up session. See docs.")
+
+    return True
+
+
+@asyncio.coroutine
+def _setup_deconz(hass, config, deconz_config):
+    """Setup Deconz session.
+
+    Load config data for server information.
+    Load light data to know which lights are available.
+    Load sensor data to know which sensors are available.
+    Start websocket for push notification of state changes from Deconz.
+    """
+    from pydeconz import DeconzSession
+    deconz = DeconzSession(hass.loop, **deconz_config)
+    hass.data[DECONZ_DATA] = deconz
+    yield from deconz.populate_config()
+    yield from deconz.populate_groups()
+    yield from deconz.populate_lights()
+    yield from deconz.populate_sensors()
+    hass.async_add_job(discovery.async_load_platform(hass,
+                                                     'light',
+                                                     DOMAIN,
+                                                     {},
+                                                     config))
+    hass.async_add_job(discovery.async_load_platform(hass,
+                                                     'binary_sensor',
+                                                     DOMAIN,
+                                                     {},
+                                                     config))
+    hass.async_add_job(discovery.async_load_platform(hass,
+                                                     'sensor',
+                                                     DOMAIN,
+                                                     deconz_config,
+                                                     config))
+    deconz.start()
+
+
+class DeconzEvent(object):
+    """When you want signals instead of entities."""
```

Comment:
I'm afraid I don't understand what that means. :(

Reply:
I understand this since you only had time to spend on this file ;).

It is a event class that will take a device sensor and generate events in hass instead of creating a sensor entity.

I will improve the pydocstring.